src/network/clientpackethandler.cpp

```diff
@@ -982,6 +982,22 @@ void Client::handleCommand_SpawnParticle(NetworkPacket* pkt)
 	m_client_event_queue.push(event);
 }
 
+/* this function should not exist. alas, i managed to leave out critical fields
+ * in the original tween serialization/deserialization code and now the protocol
+ * is fixed, so now we all have to suffer for my abject stupidity. j o y */
+template <typename T>
+inline bool readMissingTweenParamsFor(std::istringstream& is, ParticleParamTypes::TweenedParameter<T>& t) {
```

Comment:
Suggestion: bump the protocol version and integrate these read/write instructions into the existing `serialize`/ `deSerialize` functions. Benefits:

 * Cleaner code structure
 * Easier maintenance (and legacy code drop) in the future

Reply:
bumping the protocol version will only make the code messier. it won't negate the need to account for older clients; it will just create an extra code path to maintain. the possibility of eventually dropping the legacy code is the only reason i see for doing that, and i'm skeptical that will actually ever happen.

> No packet-special handling needed in every place where TweenedParameter's are exchanged between server <--> client

er, i'm not sure you understand what's going on here? the only special-casing is for the legacy parameters. all new tweenedparameters are marshalled directly with their own `de/Serialize` member functions. no additional calls to the inline functions i added should EVER be written (unless someone else eventually fucks up in precisely the same way i did, i suppose)

the new read-write functions do not belong in the `TweenedParameter` namespace. they exist only to be used in the files in which they are defined, for one very specific purpose. that's why they're marked inline. if these functions are ever called anywhere besides the invocations i wrote, something has gone horribly wrong. they are an attempt to make an ugly hack more manageable, not an interface to be used going forward

Reply:
Who instructed you to avoid protocol bumps? In this case it makes perfectly sense to me because embedding the legacy parameters into existing (de)serialize functions would result in a cleaner code structure.

Reply:
> this could've been implemented much cleaner simply by bumping the protocol version. Well, now it's too late to change that.

i agree completely

> Who instructed you to avoid protocol bumps

i believe that was a conversation i had with sfan5 but i'm not 100% on that

Reply:
Might I propose to change `p.pos.end.deSerialize(is);` to `p.pos.deSerializeCompatible(is);`, implemented similar to this this?

```C++
		// For tweenable particle spawners where the old packet format already includes the start parameter
		void deSerializeCompatible(std::istream &is, u16 protocol_version)
		{
			if (protocol_version >= 42) {
				style = static_cast<TweenStyle>(readU8(is));
				reps = readU16(is);
				beginning = readF32(is);
			}
			// Start is already serialized earlier
			end.deSerialize(is);
		}
```

This is to avoid further clutter if there's anything added in the future.

Reply:
also i'm having a hard time figuring out the purpose of this suggestion? as long as all new fields use the appropriate serialize and deserialize member functions for their parameter class, there will be no "clutter". the clutter that currently exists is due to the need to preserve interoperability with a legacy format; it's not like new legacy fields are going to unexpectedly manifest some day

Reply:
In case of adding more parameters in the future, it would provide a saner basis to work on (rather than continuing the patchwork). However, if you think it's is somewhat feature-complete, the proposed changes fit.

My suggestion was towards combining/unifying the functionality into actual (de)serialization functions rather than relying on helper functions that are spread across several files.

Reply:
i've said this like three times already but the serialize/deSerialize functions *already do this*. there is no behavior to add